### ISSUE TITLE
Port the program to HIP 

### ISSUE THREAD

**Issue by jinz2014** [cuhk-eda/CULS#7]
Hi,

I tried to port the CUDA program to HIP. On AMD GPUs, the warpSize is 64.  Could you please point out the code changes required for your large codebase ?

For example, I found the code
```
template <int STACK_SIZE>
__global__ void getCutTruthTableConsecutive(const int * pFanin0, const int * pFanin1, const int * vNumSaved,
                                            const int * vIndices, const int * vCuts, const int * vCutRanges,
                                            unsigned * vTruth, const int * vTruthRanges, const unsigned * vTruthElem,
                                            int nIndices, int nPIs, int nMaxCutSize, int * vNode2ConeResynIdx = NULL) {
    typedef cub::WarpScan<int> WarpScan;
    __shared__ typename WarpScan::TempStorage temp_storage[THREAD_PER_BLOCK / 32];
    __shared__ unsigned * vTruthMemAlloc[THREAD_PER_BLOCK / 32];

    int nThreads = NUM_BLOCKS(nIndices, 32) * 32;
    int idx = blockIdx.x * blockDim.x + threadIdx.x;
    int warpIdx = threadIdx.x / 32, laneIdx = threadIdx.x % 32;
    int stack[STACK_SIZE], stackRes[STACK_SIZE], visited[STACK_SIZE];
 
  assert(nIndices <= nThreads && nThreads - nIndices < 32);
```
Is it right that all "32" values should be replace with 64 ?

Thanks

**Comment by tefantasy** [comment]
For the case you have shown, you can replace 32 with 64. 

We do not use warp-level parallelism very often in this project, so it is possible that the case you have shown is the only one. If you have further encountered warp-level code snippets that you are not sure of, you can publish in this issue and I will take a look.

**Comment by jinz2014** [comment]
Okay. I made the changes. However, running the program shows assertion errors. I have no idea of the causes.  

./main -c "read ./benchmarks/arithmetic/hyp.aig; resyn2; write hyp_resyn2.aig"


refactor: perform MFFC covering algorithm
Refactor: setting hipLimitStackSize = 65536
Refactor: checked hipLimitStackSize = 65536
Gathered 128 POs
...
...
Level 78, global list len 1149
Level 79, global list len 270
Level 80, global list len 28
Level 81, global list len 0

Total number of cones to be resyned: 51560
Truth table computation time: 0.00 sec

refactor_core.cu:287: void factorFromTruth(const int *, const int *, uint64 *, int *, int *, int *, const unsigned int *, const unsigned int *, const int *, const unsigned int *, int): Device-side assertion `vSubgLens[currRowIdx] > 0' failed.


./alg_factor.cuh:175: int sop::sopFactorTrivialCubeRec(unsigned int, int, int, subgUtil::Subg<SUBG_CAP> *): Device-side assertion `uCube' failed.

./alg_factor.cuh:183: int sop::sopFactorTrivialCubeRec(unsigned int, int, int, subgUtil::Subg<SUBG_CAP> *): Device-side assertion `iLit != -1' failed

**Comment by tefantasy** [comment]
If I understand correctly, the error was triggered when you ran the translated HIP code, is that right? 

May I know whether you have tried running our original CUDA code on a NVIDIA GPU? I cannot reproduce the error with our hardware.

**Comment by jinz2014** [comment]
Yes, it was triggered by the HIP code on an AMD GPU only.  I think your domain knowledge may shed some light on potential causes.

**Comment by tefantasy** [comment]
It is difficult to find out what goes wrong by looking at the log only. However, one possible error in factorFromTruth is due to insufficient stack size for recursive device function calls. Although I could see from your log that the stack size was successfully set to 65536 which is sufficient on our NVIDIA GPU, I am not sure whether this is the case on an AMD GPU. Besides, note that the error(s) might also reside in previous kernels, which produce wrong data but do not trigger any assertion until factorFromTruth.
